### PR TITLE
Add `put_diagonal/3`

### LINKED ISSUE #817: Functions set_diagonal(tensor, opts \\ []) , and indexed_set(target, indices, updates)
It would be great if `set_diagonal(tensor, opts \\ []) ` and ` indexed_set(target, indices, updates) ` can be added. 

It would complement the operations allowed already. Slicing and access syntax allow setting the values to columns or rows, or individual "cells". 

Setting diagonal or other set of "cells" can be now done only by some cycle accessing and setting each diagonal value. 
There are functions that can be used as a head start: 
1) Accesign diagonal: `make_diagonal(tensor, opts \\ [])`, `take_diagonal(tensor, opts \\ [])`.
2) Adding value to the enumerated list of indexes: ` indexed_add(target, indices, updates)`.

For setting the diagonal the workaround is to take the diagonal values and subtract them from matrix, e.g. 
```
#make diagonal zero and add new value
t = Nx.random_uniform({5,5},1,0.1)
new_diagonal = Nx.random_uniform({5},1, 0.1) 
s = t |> Nx.take_diagonal() |> Nx.make_diagonal()
Nx.subtract(t, s) |> Nx.add(Nx.make_diagonal(new_diagonal))
```
The issue I see with this is that subtracting leads to rounding errors, and efficiency of the workaround.

The function ` indexed_set(target, indices, updates) ` has the similar logic and would be great complement to `indexed_add(target, indices, updates)`.

### PR DESCRIPTION
**Pull request by Benjamin-Philip**
Closes #817.

### CODE REVIEW

**Comment by Benjamin-Philip** [comment]
Let me know if this API looks okay.

File: nx/lib/nx.ex [hunk 1]
```diff
@@ -1505,7 +1505,154 @@ defmodule Nx do
...
+
+      iex> Nx.put_diagonal(Nx.iota({3, 3}), Nx.iota({3}), offset: 4)
+      ** (ArgumentError) offset must be less than length of axis 1 when positive, got: 4
+
+      iex> Nx.put_diagonal(Nx.iota({3, 3}), Nx.iota({3}), offset: -3)
+      ** (ArgumentError) absolute value of offset must be less than length of axis 0 when negative, got: -3
+  """
+  @doc type: :creation
+  def put_diagonal(tensor, diagonal, opts \\ [offset: 0]) do
+    shape = tensor |> to_tensor |> Map.get(:shape)
```
Comment (josevalim) [inline nx/lib/nx.ex:1648]:
```suggestion
    %{shape: shape} = tensor = to_tensor(tensor)
```

File: nx/lib/nx.ex [hunk 2]
```diff
@@ -1505,7 +1505,154 @@ defmodule Nx do
...
+   Given invalid offsets:
+
+      iex> Nx.put_diagonal(Nx.iota({3, 3}), Nx.iota({3}), offset: 4)
+      ** (ArgumentError) offset must be less than length of axis 1 when positive, got: 4
+
+      iex> Nx.put_diagonal(Nx.iota({3, 3}), Nx.iota({3}), offset: -3)
+      ** (ArgumentError) absolute value of offset must be less than length of axis 0 when negative, got: -3
+  """
+  @doc type: :creation
+  def put_diagonal(tensor, diagonal, opts \\ [offset: 0]) do
```
Comment (josevalim) [inline nx/lib/nx.ex:1647]:
```suggestion
  def put_diagonal(tensor, diagonal, opts \\ []) do
    opts = keyword!(opts, offset: 0)
```

The way it is currently written means someone could do this `put_diagonal(tensor, diagonal, [])` and then no default would be set.

File: nx/lib/nx.ex [hunk 3]
```diff
@@ -1505,7 +1505,154 @@ defmodule Nx do
...
+
+  @doc """
+  Puts the individual values from a 1D diagonal into the diagonal indices
+  of the given 2D tensor.
+
+  See also: `take_diagonal/2`, `make_diagonal/2`.
+
+  ## Examples
+
+     Given a 2D tensor and a 1D diagonal:
```
Comment (josevalim) [inline nx/lib/nx.ex:1519]:
```suggestion
 Given a 2D tensor and a 1D diagonal:
```

File: nx/lib/nx.ex [hunk 4]
```diff
@@ -1505,7 +1505,154 @@ defmodule Nx do
...
+        s64[4][3]
+        [
+          [1, 0, 0],
+          [0, 2, 0],
+          [0, 0, 3],
+          [0, 0, 0]
+        ]
+      >
+
+    Given a 2D tensor and a 1D diagonal with a positive offset:
```
Comment (josevalim) [inline nx/lib/nx.ex:1563]:
```suggestion
  Given a 2D tensor and a 1D diagonal with a positive offset:
```

File: nx/lib/nx.ex [hunk 5]
```diff
@@ -1505,7 +1505,154 @@ defmodule Nx do
...
+        s64[4][3]
+        [
+          [0, 1, 0],
+          [0, 0, 2],
+          [0, 0, 0],
+          [0, 0, 0]
+        ]
+      >
+
+    Given a 2D tensor and a 1D diagonal with a negative offset:
```
Comment (josevalim) [inline nx/lib/nx.ex:1587]:
```suggestion
  Given a 2D tensor and a 1D diagonal with a negative offset:
```

File: nx/lib/nx.ex [hunk 6]
```diff
@@ -1505,7 +1505,154 @@ defmodule Nx do
...
+
+
+  ## Options
+
+    * `:offset` - offset used for putting the diagonal.
+      Use offset > 0 for diagonals above the main diagonal,
+      and offset < 0 for diagonals below the main diagonal.
+      Defaults to 0.
+
+
```
Comment (josevalim) [inline nx/lib/nx.ex:1619]:
```suggestion
```

File: nx/lib/nx.ex [hunk 7]
```diff
@@ -1505,7 +1505,154 @@ defmodule Nx do
...
+      iex> Nx.put_diagonal(Nx.iota({3, 3}), Nx.iota({3, 3}))
+      ** (ArgumentError) put_diagonal/3 expects diagonal of rank 1, got tensor of rank: 2
+
+      iex> Nx.put_diagonal(Nx.iota({3, 3}), Nx.iota({2}))
+      ** (ArgumentError) expected diagonal tensor of length: 3, got diagonal tensor of length: 2
+
+      iex> Nx.put_diagonal(Nx.iota({3, 3}), Nx.iota({3}), offset: 1)
+      ** (ArgumentError) expected diagonal tensor of length: 2, got diagonal tensor of length: 3
+
+   Given invalid offsets:
```
Comment (josevalim) [inline nx/lib/nx.ex:1638]:
```suggestion
  Given an invalid tensor:

      iex> Nx.put_diagonal(Nx.iota({3, 3, 3}), Nx.iota({3}))
      ** (ArgumentError) put_diagonal/3 expects tensor of rank 2, got tensor of rank: 3

  Given invalid diagonals:

      iex> Nx.put_diagonal(Nx.iota({3, 3}), Nx.iota({3, 3}))
      ** (ArgumentError) put_diagonal/3 expects diagonal of rank 1, got tensor of rank: 2

      iex> Nx.put_diagonal(Nx.iota({3, 3}), Nx.iota({2}))
      ** (ArgumentError) expected diagonal tensor of length: 3, got diagonal tensor of length: 2

      iex> Nx.put_diagonal(Nx.iota({3, 3}), Nx.iota({3}), offset: 1)
      ** (ArgumentError) expected diagonal tensor of length: 2, got diagonal tensor of length: 3

  Given invalid offsets:
```

**Review by josevalim — COMMENTED** [review]
Just some minor comments and this looks good to go. WDYT @polvalente?

File: nx/lib/nx.ex (hunk 2 above)
Comment (Benjamin-Philip) [reply nx/lib/nx.ex:1647]: In this case

```elixir
offset = opts[:offset] || 0
```

should be good enough, right?

Comment (josevalim) [reply nx/lib/nx.ex]: Let's use `keyword!` as it also validates against invalid options.

File: nx/lib/nx.ex (hunk 7 above)
Comment (Benjamin-Philip) [reply nx/lib/nx.ex:1638]: Decreasing the indentation by 2 spaces actually removes the doctests and results in invalid rendering of the docs.

Comment (josevalim) [reply nx/lib/nx.ex:1638]: I removed the indentation only of the text, not of the code. :)

File: nx/lib/nx.ex [hunk 8]
```diff
@@ -1505,7 +1505,152 @@ defmodule Nx do
...
+      iex> Nx.put_diagonal(Nx.iota({3, 3}), Nx.iota({3}), offset: 4)
+      ** (ArgumentError) offset must be less than length of axis 1 when positive, got: 4
+
+      iex> Nx.put_diagonal(Nx.iota({3, 3}), Nx.iota({3}), offset: -3)
+      ** (ArgumentError) absolute value of offset must be less than length of axis 0 when negative, got: -3
+  """
+  @doc type: :creation
+  def put_diagonal(tensor, diagonal, opts \\ []) do
+    %{shape: shape} = tensor = to_tensor(tensor)
+    offset = opts |> keyword!(offset: 0) |> Keyword.get(:offset)
```
Comment (josevalim) [inline nx/lib/nx.ex:1649]:
```suggestion
    offset = opts |> keyword!(offset: 0) |> Keyword.fetch!(:offset)
```

File: nx/lib/nx/shape.ex [hunk 9]
```diff
@@ -1269,6 +1269,93 @@ defmodule Nx.Shape do
           "make_diagonal/2 expects tensor of rank 1, got tensor of rank: #{tuple_size(invalid_shape)}"
   end
 
+  @doc """
+  Validates the tensor, diagonal and offset given to Nx.put_diagonal/3.
```
Comment (polvalente) [inline nx/lib/nx/shape.ex:1273]:
```suggestion
  Validates the tensor, diagonal and offset given to `Nx.put_diagonal/3`.
```

**Review by polvalente — COMMENTED** [review]
Looks good to me!

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**